### DISCUSSION TITLE
How to register a recipe helper script from the taxonomies module

### DISCUSSION THREAD

**Discussion by giannik** [OrchardCMS/OrchardCore#11488] [Q&A]
I need to register a recipe helper script to call from a recipe like below 

```
   "ProductCatalogItem": {
      "ContentItemIds": "[js: taxonomyTermIdByAlias('alias:free-trial')]"
    },
```

Looking at other code examples  like in CommonGeneratorMethods.cs  seems that you need access to  OrchardCoreBuilder class in Startup

But in Taxonomys module I try this 

```
  public class Startup : StartupBase
    {
  public override   void ConfigureServices(IServiceCollection services)
        {
             var builder = services.AddOrchardCore();    //Im adding this to get access to OrchardCoreBuilder
             builder.AddScripting();  //call an extension method on builder to register recipe helper
       }

    }
```

But this doesnt work because it is  a static class

So my question is how to register correctly a recipe helper script to be used from a module ? In this case the taxonomies module

**Comment by lampersky** [comment]
Inside my custom module I'm registering new script method this way:
**MyGuidGenerator.cs**
```
    public class MyGuidGeneratorMethod : IGlobalMethodProvider
    {
        private static GlobalMethod MyGuid = new GlobalMethod
        {
            Name = "MyGuid",
            Method = serviceProvider => (Func<string>)(() =>
            {
                return Guid.NewGuid().ToString();
            })
        };

        public IEnumerable<GlobalMethod> GetMethods()
        {
            yield return MyGuid;
        }
    }
```

module **Startup.cs**:

```
    public class Startup : StartupBase
    {
        public override void ConfigureServices(IServiceCollection services)
        {
                services.AddSingleton<IGlobalMethodProvider, MyGuidGeneratorMethod>();
        }
    }
```

and now you can use this new scripting method for example within workflow:

![image](https://user-images.githubusercontent.com/6403130/161599828-543937e5-864c-44ba-8982-4d6844d962ba.png)

or inside a recipe:

![image](https://user-images.githubusercontent.com/6403130/161601127-85e582d6-ade0-4c5e-8bdb-576ad81ba6a6.png)

**Comment by giannik** [comment]
I didi the exact think with my method , only difference is Im passing parameter : 

```
    public class CommonTaxonomyMethods : IGlobalMethodProvider
    {

        /// <summary>
        /// Returns the term id given a taxonomy and term alias
        /// example  alias:taxonomyterm
        /// </summary>
        private static GlobalMethod TaxonomyIdByAlias = new GlobalMethod
        {
            Name = "taxonomyTermIdByAlias",
            Method = serviceProvider => (Func<string, string>)  ( termAlias =>
            {
                var handleManager = serviceProvider.GetRequiredService<IContentHandleManager>();
                var taxonomyTermContentItemId =   handleManager.GetContentItemIdAsync(termAlias).Result;
                return taxonomyTermContentItemId;

            })
        };


        public IEnumerable<GlobalMethod> GetMethods()
        {
            yield return TaxonomyIdByAlias;;
        }
    }
```


But puting a breakpoint on the method and running it from a workflow it doesnt get hit.

**Reply by lampersky** [reply]
please show us, how you registered CommonTaxonomyMethods (DI), other than that have you enabled your custom module (feature)?

**Reply by lampersky** [reply]
also change **.Result;** to **.GetAwaiter().GetResult();**

**Reply by giannik** [reply]
Ill try. 
Thing is Im putting a breakpoint in method body and its not getting hit.

**Reply by lampersky** [reply]
I've just added parameter to the method, setup breakpoint, and as you can see:
![image](https://user-images.githubusercontent.com/6403130/161607374-a79574ac-5ee7-40ac-b7e5-e48144d3002c.png)
everything works

**Comment by giannik** [comment]
In the taxonomies module Im registering :

```
 services.AddSingleton<IGlobalMethodProvider, CommonTaxonomyMethods>();
```

and from a CreateContent activity in woklow I run the following json :
```
 {

  "Owner": "{{Workflow.Input.User.UserId}}",
  "Author": "{{Workflow.Input.User.UserName}}",
  "Subscription": {
    "StartDate": {
      "Value": "2021-04-19T00:00:00Z"
    },
    "EndDate": {
      "Value": "2021-04-30T00:00:00Z"
    },
    "ProductCatalogItem": {
      "ContentItemIds": ["[js:taxonomyTermIdByAlias('alias:free-trial')]"]
    },
    "Customer": {
      "ContentItemIds": [
        "{{Workflow.Properties.NewCustomerContentItem.ContentItemId}}"
      ]
    }
  }
  
}
```

**Reply by lampersky** [reply]
so you are working with **Taxonomies module** source code and trying to tweak it a bit, by adding some additional scripting helpers, is that module activated?
![image](https://user-images.githubusercontent.com/6403130/161608281-cd05a722-8279-4f6e-84bb-fbe680a0733d.png)

**Reply by giannik** [reply]
yes, its enabled

**Reply by lampersky** [reply]
share your taxonomy **Startup.cs** so we can have a look, may I ask why you are modifying existing module? maybe it would be better to leave Taxonomy as is, and create custom feature with all those new methods which you need

**Reply by giannik** [reply]
sure Ill place it as separate feature, now I just want to see it works.
Im thinking something to do with the scripting module not picking it as a script . Will do some further debugging.
Thanks

**Reply by giannik** [reply]
at least I know Im not doing something wrong

**Reply by giannik** [reply]
@lampersky  seems that the CreateContent task uses the liquid expression evaluation 
while the example you did with script task uses the scripting evaluator.
So I guess for the createcontent task you can only use liquid , no js  expressions !

**Reply by lampersky** [reply]
@giannik please note **ContentItemIds** is an array, in your very first example you are trying to assign string, please try sth like this:

```
"ContentItemIds": [
  "[js:MyGuid('test')]",
  "[js:MyGuid('test')]",
  "[js:MyGuid('test')]"
]
```

**Reply by giannik** [reply]
yes Im of that.
Thanks
